### ISSUE TITLE
Missing column crash after migration is performed

### ISSUE THREAD

**Issue by 0xpablo** [groue/GRDB.swift#1453]
### What did you do?

We are noticing that after we perform a migration and begin a database observation right after, some users face a DatabaseValueConversion error like this one: `DatabaseValueConversion.swift:160: Fatal error: could not read Bool from missing column new_column`.
What I'm finding surprising is that the error also prints out the row, and you can see the supposedly missing column in the row.
```
(row: [id:"ID" new_column:0 ...])
```
We have double checked that the migrations are ran before starting the subscription (we are not doing it in parallel or anything like that).

For migrations, we are using `DatabaseMigrator` like this:
```swift
    var migrator = DatabaseMigrator()
    migrator.registerMigration(Migration.newColumn.rawValue) { db in
        try db.alter(table: Table.databaseTableName) { t in
            t.add(column: Table.Fields.newColumn, .boolean).defaults(to: false)
        }
    }
    try migrator.migrate(database)
```

We are finding it very hard to reproduce the problem and it seems to be affecting ~10% of our users after we do a migration.
I would like to find time to upgrade to a newer version, but I wanted to ask if this is a known issue of a previous version or if this is something that can happen and we need to tweak the config do anything else to guarantee that this does not happen. Wondering if `DatabaseSchemaCache` could be involved in this issue.

Thanks a lot in advance! And thanks for this awesome library 🙂

### What did you expect to happen?

That the app did not crash because the migration was performed.

### What happened instead?

The app crashed with a fatal error 

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 4.14.0
**Installation method:** SPM
**Xcode version:** 14.3.1
**Swift version:** 5.8
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** macOS 13.6

**Comment by groue** [comment]
Hello @0xpablo,

> Wondering if DatabaseSchemaCache could be involved in this issue.

Good question.

Conversions are independent from the schema cache. We just decode whatever we find in the fetched row. So if `new_column` is missing, it is indeed missing.

But your debugging output clearly shows the missing column: `row: [id:"ID" new_column:0 ...]` 😅 

A possibility is that your app decodes a request that involves associations (`including(required:)`, etc.) Such requests do indeed use the schema cache, in order to determine how many columns belong to each one of the fetched records:

```swift
// WHAT HAPPENS WHEN WE DECODE A REQUEST WITH ASSOCIATIONS:
//
// SELECT player.*, team.* ...
// id: 1, name: "Arthur", score: 1000, id: 1, name: "Reds"
// <-------------------------------->  <----------------->
//         3 player columns              2 team columns
Player.including(required: Player.team)
```

If the cache is wrong, and tells that player has only two columns, then surprising things will happen. In the example below, the `Player` calls will complain about a missing `score` column, and `Team` will complain about a missing `name` column, even though the row contains all the needed information:

```swift
// WHAT HAPPENS WHEN WE DECODE A REQUEST WITH ASSOCIATIONS
// AND A BOGUS SCHEMA CACHE THAT THINKS THE player TABLE
// ONLY HAS TWO COLUMNS:
//
// SELECT player.*, team.* ...
// id: 1, name: "Arthur", score: 1000, id: 1, name: "Reds"
// <------------------->  <----------------->
//    2 player columns      2 team columns
Player.including(required: Player.team)
```

This really looks like your issue. If you confirm that the error happens when decoding such a request, then we may have a clue.

---

Generally speaking, the GRDB schema cache is intended to never contain stale information, **as long as** the database schema is only modified from the one and single instance of `DatabaseQueue` or `DatabasePool` that deals with that database in the app.

The precondition is important:

- If the database schema is modified from another process, you may need to call [`Database.clearSchemaCache()`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/database/clearschemacache()).
- If the database schema is modified by another instance of `DatabaseQueue` or `DatabasePool` within the same process, you may need to call [`Database.clearSchemaCache()`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/database/clearschemacache()), but mainly to look at the [Rule 1 of GRBD Concurrency](https://swiftpackageindex.com/groue/grdb.swift/v6.21.0/documentation/grdb/concurrency#Rule-1-Connect-to-any-database-file-only-once).

As long as the database schema is handled by a single instance of `DatabaseQueue` or `DatabasePool`, you should never have to think about the schema cache. Any bug related to this cache is a GRDB bug that needs to be fixed, not an application bug.

Now, the schema cache has continuously been tweaked since GRDB 4.14.0 (April 2020), in order to support temporary tables, views, attached databases, etc. Some bugs have probably been fixed, but I don't have a clear memory of all the changes performed in more than three years.

---

> We are noticing that after we perform a migration and begin a database observation right after [...]

`ValueObservation` is another area that has been continuously improved in the recent years, with *many* improvements and bug fixes.

---

> I would like to find time to upgrade to a newer version.

You might find it easier if you do it in two steps:

1. Migrate to GRDB 5: https://github.com/groue/GRDB.swift/blob/master/Documentation/GRDB5MigrationGuide.md
2. Migrate to GRDB 6: https://github.com/groue/GRDB.swift/blob/master/Documentation/GRDB6MigrationGuide.md

I know that such upgrade can take some time, so it might be worth seeing in the issue can be solved from within your app (at constant GRDB version).

So @0xpablo, please report:

1. If your app do use a single `DatabaseQueue` or `DatabasePool`.
2. What you mean precisely by "perform a migration and begin a database observation right after", with some sample code.

**Comment by groue** [comment]
Thinking a little bit more about it:

> So @0xpablo, please report:
> 
> 1. If your app do use a single `DatabaseQueue` or `DatabasePool`.
> 2. What you mean precisely by "perform a migration and begin a database observation right after", with some sample code.

The first point remains very important. It is a misuse to open multiple connections to single database file. In other words, the eventual bad consequences are an application bug, not a GRDB bug.

The second point remains interesting, but not very relevant unless you share some really bizarre code 😅

You might have good results if you set the [`requiresWriteAccess`](https://github.com/groue/GRDB.swift/tree/v4.14.0#valueobservationrequireswriteaccess) flag on your `ValueObservation`. This will force `ValueObservation` to fetch values from the writer connection, which can't be suspected of having a stale schema cache. This is a workaround for GRDB 4 only: this flag will need to be removed when you upgrade to GRDB 6.

**Comment by 0xpablo** [comment]
Hi @groue, thanks a lot for your detailed reply and all of those resources, it's much appreciated!

The associations part is interesting, I took a look at our observation and while we don't use the `including` API because we build the query ourselves, we have quite a bunch of `LEFT JOIN` statements in this case. I suppose the `including` API uses `JOIN` behind the scenes so this might be related.

I will definitely make some time to migrate to GRDB 6 soon. I have been looking into the guides for a few days and I think we are ready to upgrade. That said, do you think that using `clearSchemaCache()` and/or checkpointing the database could help in the meantime?

> If your app do use a single DatabaseQueue or DatabasePool.

We only use a single  `DatabaseQueue`.

> What you mean precisely by "perform a migration and begin a database observation right after", with some sample code.

I mean that I made sure that we wait for the migration to finish to issue the subscriptions. We guard DB initialization status and we don't transition to the initialized status until we run the migrations.
Code that starts the observations listens to the initialization status to begin them, the `ValueObservation` is not created nor started until our database has been initialized.

Thanks again for your support

**Comment by groue** [comment]
Yes, @0xpablo, please report if the fix solves the issue. Your answer may help other GRDB users!

Also, I'm curious about the operating system version(s) of the users who meet the issue. Could you please give this information, if you happen to have it? I ask because SQLite has not always been perfect about reporting schema changes (and tell GRDB to invalidate its schema cache). Several workarounds have been applied to GRDB over time, as holes were discovered. But there may still exist some!

OK, I'll close the issue but feel free to reopen if the fix doesn't work as expected!

And maybe start planning about an upgrade of your GRDB copy 😅 I'm already planning for GRDB 7, because there are screws to tighten regarding Swift concurrency (async/await, Sendable, etc)!

**Comment by 0xpablo** [comment]
Hi @groue , The issue seems to be affecting all iOS versions we support (iOS 15, 16 and 17).
We are beta-testing the changes with the manual `clearSchemaCache` as our last initialization step but saw one crash during beta testing, so unfortunately it seems that the workaround is not working.

We've investigated a bit more and found something interesting:

While most of the crashes show the field for the migration that was recently added, we also found cases of our very first migration, a migration from 2018. It's very unlikely (I'd say impossible) that the user hasn't migrated until now.

Let me attach a couple of stack traces in case it helps. I left the threads with relevant GRDB work going on.

I'll probably allocate some time next week to review our DB code in case we are violating any API contract and also  perform the GRDB upgrade as we are running out of ideas.

Thanks again!

Crash 1:
```
rashed: GRDB.DatabasePool.reader.2
0  libswiftCore.dylib             0x3886c _assertionFailure(_:_:file:line:flags:) + 264
1  Goodnotes                      0x20b5ce8 fatalConversionError<A>(to:from:conversionContext:file:line:) + 141 (DatabaseValueConversion.swift:141)
2  Goodnotes                      0x20b6e20 partial apply for implicit closure #1 in static DatabaseValueConvertible.decodeIfPresent(from:atUncheckedIndex:) + 4366708256 (<compiler-generated>:4366708256)
3  Goodnotes                      0x20b5d18 fatalConversionError<A>(to:from:in:atColumn:file:line:) + 4366703896 (<compiler-generated>:4366703896)
4  Goodnotes                      0x1ce18c8 specialized FolderToFolderItem.init(row:) + 48 (FolderToFolderItem +GRDB.swift:48)
5  Goodnotes                      0x1cdfc8c protocol witness for FetchableRecord.init(row:) in conformance FolderToFolderItem + 4362681484 (<compiler-generated>:4362681484)
6  Goodnotes                      0x217d63c RecordCursor.next() + 456 (FetchableRecord.swift:456)
7  Goodnotes                      0x217e338 protocol witness for Cursor.next() in conformance RecordCursor<A> + 4367524664 (<compiler-generated>:4367524664)
8  Goodnotes                      0x20719a0 specialized Array.init<A>(_:) + 22 (Cursor.swift:22)
9  Goodnotes                      0x217d514 static FetchableRecord.fetchAll(_:arguments:adapter:) + 189 (FetchableRecord.swift:189)
10 Goodnotes                      0x217d86c static FetchableRecord.fetchAll<A>(_:_:) + 334 (FetchableRecord.swift:334)
11 Goodnotes                      0x1e274fc specialized closure #1 in closure #1 in Reactive<A>.fetchAll(in:) + 4364023036 (<compiler-generated>:4364023036)
12 Goodnotes                      0x1e288cc partial apply for specialized closure #1 in closure #1 in Reactive<A>.fetchAll(in:) + 4364028108 (<compiler-generated>:4364028108)
13 Goodnotes                      0x209dc10 closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 376 (DatabasePool.swift:376)
14 Goodnotes                      0x20a2d4c partial apply for closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4366626124 (<compiler-generated>:4366626124)
15 Goodnotes                      0x208ff7c Database.inTransaction(_:_:) + 821 (Database.swift:821)
16 Goodnotes                      0x209db10 closure #1 in closure #1 in DatabasePool.read<A>(_:) + 373 (DatabasePool.swift:373)
17 Goodnotes                      0x20a2d2c partial apply for closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4366626092 (<compiler-generated>:4366626092)
18 Goodnotes                      0x20e1a30 closure #1 in SerializedDatabase.sync<A>(_:) + 4366883376
19 Goodnotes                      0x20e27b8 partial apply for closure #1 in SerializedDatabase.reentrantSync<A>(_:) + 4366886840
20 Goodnotes                      0x20e2864 partial apply for closure #1 in SerializedDatabase.sync<A>(_:) + 4366887012
21 libswiftDispatch.dylib         0x1594 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
22 libswiftDispatch.dylib         0x1570 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
23 libswiftDispatch.dylib         0x14b0 closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
24 libswiftDispatch.dylib         0x13e8 partial apply for thunk for @callee_guaranteed () -> () + 28
25 libswiftDispatch.dylib         0x13c0 thunk for @escaping @callee_guaranteed () -> () + 28
26 libdispatch.dylib              0x4300 _dispatch_client_callout + 20
27 libdispatch.dylib              0x136b4 _dispatch_lane_barrier_sync_invoke_and_complete + 56
28 libswiftDispatch.dylib         0x2078 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.asyncAndWait<A>(execute:) + 192
29 libswiftDispatch.dylib         0x1fb0 partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 76
30 libswiftDispatch.dylib         0x1e30 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
31 libswiftDispatch.dylib         0x1c74 OS_dispatch_queue.asyncAndWait<A>(execute:) + 140
32 libswiftDispatch.dylib         0x1be0 OS_dispatch_queue.sync<A>(execute:) + 64
33 Goodnotes                      0x20e09a8 SerializedDatabase.sync<A>(_:) + 96 (SerializedDatabase.swift:96)
34 Goodnotes                      0x20a23cc partial apply for closure #1 in DatabasePool.read<A>(_:) + 4366623692
35 Goodnotes                      0x20a22cc partial apply for closure #1 in DatabasePool.read<A>(_:) + 4366623436
36 Goodnotes                      0x21a1f44 Pool.get<A>(block:) + 95 (Pool.swift:95)
37 Goodnotes                      0x209e17c DatabasePool.read<A>(_:) + 4366606716
38 Goodnotes                      0x209d678 DatabasePool.read<A>(_:) + 4366603896
39 Goodnotes                      0x209f024 protocol witness for DatabaseReader.read<A>(_:) in conformance DatabasePool + 4366610468 (<compiler-generated>:4366610468)
40 Goodnotes                      0x1d97b74 LoggingDatabaseWriter.read<A>(_:) + 127 (LoggingDatabaseWriter.swift:127)
41 Goodnotes                      0x1d98400 protocol witness for DatabaseReader.read<A>(_:) in conformance LoggingDatabaseWriter + 4363437056 (<compiler-generated>:4363437056)
42 Goodnotes                      0x1e23130 specialized closure #1 in Reactive<A>.fetchAll(in:) + 638 (LibraryManager.swift:638)
43 Goodnotes                      0x1ad7ae0 closure #1 in static PrimitiveSequenceType<>.create(subscribe:) + 32 (Single.swift:32)
44 Goodnotes                      0x1a8f784 AnonymousObservableSink.run(_:) + 60 (Create.swift:60)
45 Goodnotes                      0x1a8f928 AnonymousObservable.run<A>(_:cancel:) + 76 (Create.swift:76)
46 Goodnotes                      0x1ac402c closure #1 in Producer.subscribe<A>(_:) + 26 (Producer.swift:26)
47 Goodnotes                      0x1ac3740 specialized CurrentThreadScheduler.schedule<A>(_:action:) + 18 (CurrentThreadScheduler.swift:18)
48 Goodnotes                      0x1ac3640 Producer.subscribe<A>(_:) + 24 (Producer.swift:24)
49 Goodnotes                      0x1ab8180 protocol witness for ObservableType.subscribe<A>(_:) in conformance Observable<A> + 4360421760 (<compiler-generated>:4360421760)
50 Goodnotes                      0x1add818 closure #1 in SubscribeOnSink.run() + 79 (SubscribeOn.swift:79)
51 Goodnotes                      0x1a992e0 closure #1 in DispatchQueueConfiguration.schedule<A>(_:action:) + 27 (DispatchQueueConfiguration.swift:27)
52 Goodnotes                      0x1a8f264 thunk for @escaping @callee_guaranteed () -> () + 4360254052 (<compiler-generated>:4360254052)
53 libdispatch.dylib              0x26a8 _dispatch_call_block_and_release + 32
54 libdispatch.dylib              0x4300 _dispatch_client_callout + 20
55 libdispatch.dylib              0x77b8 _dispatch_continuation_pop + 600
56 libdispatch.dylib              0x6dd4 _dispatch_async_redirect_invoke + 584
57 libdispatch.dylib              0x15be4 _dispatch_root_queue_drain + 392
58 libdispatch.dylib              0x163ec _dispatch_worker_thread2 + 156
59 libsystem_pthread.dylib        0x1928 _pthread_wqthread + 228
60 libsystem_pthread.dylib        0x1a04 start_wqthread + 8

GRDB.DatabasePool.reader.1
0  libsystem_kernel.dylib         0x1d34 pread + 8
1  libsqlite3.dylib               0x8f3c0 sqlite3_sourceid + 7688
2  libsqlite3.dylib               0x7354 sqlite3_snprintf + 7720
3  libsqlite3.dylib               0x1a098 sqlite3_log + 3480
4  libsqlite3.dylib               0x98dd0 sqlite3_sourceid + 47128
5  libsqlite3.dylib               0xa448c sqlite3_sourceid + 93908
6  libsqlite3.dylib               0xa5700 sqlite3_sourceid + 98632
7  libsqlite3.dylib               0x33e84 sqlite3_step + 35524
8  libsqlite3.dylib               0x2b798 sqlite3_step + 984
9  Goodnotes                      0x217d5e4 RecordCursor.next() + 451 (FetchableRecord.swift:451)
10 Goodnotes                      0x217e338 protocol witness for Cursor.next() in conformance RecordCursor<A> + 4367524664 (<compiler-generated>:4367524664)
11 Goodnotes                      0x2071a20 specialized Array.init<A>(_:) + 22 (Cursor.swift:22)
12 Goodnotes                      0x217d514 static FetchableRecord.fetchAll(_:arguments:adapter:) + 189 (FetchableRecord.swift:189)
13 Goodnotes                      0x217d86c static FetchableRecord.fetchAll<A>(_:_:) + 334 (FetchableRecord.swift:334)
14 Goodnotes                      0x217d304 FetchRequest<>.fetchAll(_:) + 397 (FetchableRecord.swift:397)
15 Goodnotes                      0x217c27c static FetchableRecord<>.fetchAll(_:) + 43 (FetchableRecord+TableRecord.swift:43)
16 Goodnotes                      0x1f8ef2c closure #1 in GRDBSyncStore.listStreamItems(needsSyncOnly:) + 4365496108
17 Goodnotes                      0x1f94440 partial apply for closure #1 in GRDBSyncStore.listAssetItems(needsSyncOnly:) + 4365517888
18 Goodnotes                      0x1f94408 partial apply for closure #1 in GRDBSyncStore.listStreamItems(needsSyncOnly:) + 4365517832
19 Goodnotes                      0x209dc10 closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 376 (DatabasePool.swift:376)
20 Goodnotes                      0x20a2d4c partial apply for closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4366626124 (<compiler-generated>:4366626124)
21 Goodnotes                      0x208ff7c Database.inTransaction(_:_:) + 821 (Database.swift:821)
22 Goodnotes                      0x209db10 closure #1 in closure #1 in DatabasePool.read<A>(_:) + 373 (DatabasePool.swift:373)
23 Goodnotes                      0x20a2d2c partial apply for closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4366626092 (<compiler-generated>:4366626092)
24 Goodnotes                      0x20e1a30 closure #1 in SerializedDatabase.sync<A>(_:) + 4366883376
25 Goodnotes                      0x20e27b8 partial apply for closure #1 in SerializedDatabase.reentrantSync<A>(_:) + 4366886840
26 Goodnotes                      0x20e2864 partial apply for closure #1 in SerializedDatabase.sync<A>(_:) + 4366887012
27 libswiftDispatch.dylib         0x1594 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
28 libswiftDispatch.dylib         0x1570 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
29 libswiftDispatch.dylib         0x14b0 closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
30 libswiftDispatch.dylib         0x13e8 partial apply for thunk for @callee_guaranteed () -> () + 28
31 libswiftDispatch.dylib         0x13c0 thunk for @escaping @callee_guaranteed () -> () + 28
32 libdispatch.dylib              0x4300 _dispatch_client_callout + 20
33 libdispatch.dylib              0x136b4 _dispatch_lane_barrier_sync_invoke_and_complete + 56
34 libswiftDispatch.dylib         0x2078 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.asyncAndWait<A>(execute:) + 192
35 libswiftDispatch.dylib         0x1fb0 partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 76
36 libswiftDispatch.dylib         0x1e30 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
37 libswiftDispatch.dylib         0x1c74 OS_dispatch_queue.asyncAndWait<A>(execute:) + 140
38 libswiftDispatch.dylib         0x1be0 OS_dispatch_queue.sync<A>(execute:) + 64
39 Goodnotes                      0x20e09a8 SerializedDatabase.sync<A>(_:) + 96 (SerializedDatabase.swift:96)
40 Goodnotes                      0x20a23cc partial apply for closure #1 in DatabasePool.read<A>(_:) + 4366623692
41 Goodnotes                      0x20a22cc partial apply for closure #1 in DatabasePool.read<A>(_:) + 4366623436
42 Goodnotes                      0x21a1f44 Pool.get<A>(block:) + 95 (Pool.swift:95)
43 Goodnotes                      0x209e17c DatabasePool.read<A>(_:) + 4366606716
44 Goodnotes                      0x209d678 DatabasePool.read<A>(_:) + 4366603896
45 Goodnotes                      0x209f024 protocol witness for DatabaseReader.read<A>(_:) in conformance DatabasePool + 4366610468 (<compiler-generated>:4366610468)
46 Goodnotes                      0x1f8ecdc GRDBSyncStore.listStreamItems(needsSyncOnly:) + 4365495516
47 Goodnotes                      0x1f91448 protocol witness for SyncStore.listStreamItems(needsSyncOnly:) in conformance GRDBSyncStore<A, B> + 4365505608 (<compiler-generated>:4365505608)
48 Goodnotes                      0x1c3eeb4 partial apply for specialized implicit closure #6 in implicit closure #5 in AnySyncStore.init<A>(base:) + 4362022580
49 Goodnotes                      0x1c3fb38 partial apply for specialized implicit closure #6 in implicit closure #5 in AnySyncStore.init<A>(base:) + 4362025784
50 Goodnotes                      0x1fa32b0 AnySyncStore.listStreamItems(needsSyncOnly:) + 148 (SyncStore.swift:148)
51 Goodnotes                      0x1e7c110 SyncManagerV2.listItemsToSync() + 318 (SyncManagerV2.swift:318)
52 Goodnotes                      0x1e88564 partial apply for closure #1 in SyncManagerV2.start() + 91 (SyncManagerV2.swift:91)
53 Goodnotes                      0x1e885a8 partial apply for thunk for @escaping @callee_guaranteed () -> (@owned [EventStreamIndex : (stream: SyncStreamItem<A, B>, assets: [SyncAssetItem<A, B>])], @error @owned Error) + 4364420520 (<compiler-generated>:4364420520)
54 Goodnotes                      0x22cddc8 closure #1 in closure #1 in Thenable.map<A>(on:flags:_:) + 78 (Thenable.swift:78)
55 Goodnotes                      0x22c4eac thunk for @escaping @callee_guaranteed () -> () + 4368862892 (<compiler-generated>:4368862892)
56 libdispatch.dylib              0x26a8 _dispatch_call_block_and_release + 32
57 libdispatch.dylib              0x4300 _dispatch_client_callout + 20
58 libdispatch.dylib              0xb894 _dispatch_lane_serial_drain + 748
59 libdispatch.dylib              0xc3c4 _dispatch_lane_invoke + 380
60 libdispatch.dylib              0x17004 _dispatch_root_queue_drain_deferred_wlh + 288
61 libdispatch.dylib              0x16878 _dispatch_workloop_worker_thread + 404
62 libsystem_pthread.dylib        0x1964 _pthread_wqthread + 288
63 libsystem_pthread.dylib        0x1a04 start_wqthread + 8

GRDB.DatabasePool.writer
0  libsystem_kernel.dylib         0x1d34 pread + 8
1  libsqlite3.dylib               0x8f3c0 sqlite3_sourceid + 7688
2  libsqlite3.dylib               0x7354 sqlite3_snprintf + 7720
3  libsqlite3.dylib               0x1a098 sqlite3_log + 3480
4  libsqlite3.dylib               0x98dd0 sqlite3_sourceid + 47128
5  libsqlite3.dylib               0xa448c sqlite3_sourceid + 93908
6  libsqlite3.dylib               0x380fc sqlite3_step + 52540
7  libsqlite3.dylib               0xa1570 sqlite3_sourceid + 81848
8  libsqlite3.dylib               0x2ce24 sqlite3_step + 6756
9  libsqlite3.dylib               0x2b798 sqlite3_step + 984
10 Goodnotes                      0x20d25cc specialized static Row.fetchOne(_:arguments:adapter:) + 799 (Row.swift:799)
11 Goodnotes                      0x20d27b4 specialized static Row.fetchOne<A>(_:_:) + 1058 (Row.swift:1058)
12 Goodnotes                      0x20c5b2c static Row.fetchCursor(_:sql:arguments:adapter:) + 4366768940
13 Goodnotes                      0x20c5a0c static Row.fetchOne(_:sql:arguments:adapter:) + 4366768652
14 Goodnotes                      0x1dedc90 closure #1 in LibraryDocumentsCountSubscriptionHandler.subscribe(_:onChangeHandler:onErrorHandler:) + 23 (LibraryDocumentsCountSubscriptionHandler.swift:23)
15 Goodnotes                      0x1dee0c8 partial apply for closure #1 in LibraryDocumentsCountSubscriptionHandler.subscribe(_:onChangeHandler:onErrorHandler:) + 4363788488 (<compiler-generated>:4363788488)
16 Goodnotes                      0x21d6ac8 ValueReducers.Fetch.fetch(_:) + 15 (Fetch.swift:15)
17 Goodnotes                      0x21d6b38 protocol witness for ValueReducer.fetch(_:) in conformance ValueReducers.Fetch<A> + 4367887160 (<compiler-generated>:4367887160)
18 Goodnotes                      0x21d6eb0 ValueReducers.Map.fetch(_:) + 41 (Map.swift:41)
19 Goodnotes                      0x21d704c protocol witness for ValueReducer.fetch(_:) in conformance ValueReducers.Map<A, B> + 4367888460 (<compiler-generated>:4367888460)
20 Goodnotes                      0x21d8e6c partial apply for closure #2 in ValueReducer.fetch(_:requiringWriteAccess:) + 33 (ValueReducer.swift:33)
21 Goodnotes                      0x208efe0 Database.readOnly<A>(_:) + 140 (Utils.swift:140)
22 Goodnotes                      0x21d87ec ValueReducer.fetch(_:requiringWriteAccess:) + 32 (ValueReducer.swift:32)
23 Goodnotes                      0x20bbbf0 closure #3 in DatabaseWriter.add<A>(observation:onError:onChange:) + 414 (DatabaseWriter.swift:414)
24 Goodnotes                      0x20bdae4 partial apply for closure #3 in DatabaseWriter.add<A>(observation:onError:onChange:) + 4366736100 (<compiler-generated>:4366736100)
25 Goodnotes                      0x20e258c closure #1 in SerializedDatabase.async(_:) + 185 (SerializedDatabase.swift:185)
26 Goodnotes                      0x209d1a4 thunk for @escaping @callee_guaranteed () -> () + 4366602660 (<compiler-generated>:4366602660)
27 libdispatch.dylib              0x26a8 _dispatch_call_block_and_release + 32
28 libdispatch.dylib              0x4300 _dispatch_client_callout + 20
29 libdispatch.dylib              0xb894 _dispatch_lane_serial_drain + 748
30 libdispatch.dylib              0xc3c4 _dispatch_lane_invoke + 380
31 libdispatch.dylib              0x17004 _dispatch_root_queue_drain_deferred_wlh + 288
32 libdispatch.dylib              0x16878 _dispatch_workloop_worker_thread + 404
33 libsystem_pthread.dylib        0x1964 _pthread_wqthread + 288
34 libsystem_pthread.dylib        0x1a04 start_wqthread + 8
```

Crash 2:
```

Crashed: GRDB.DatabasePool.reader.2
0  libswiftCore.dylib             0x373e4 _assertionFailure(_:_:file:line:flags:) + 264
1  Goodnotes                      0x1fb4364 fatalConversionError<A>(to:from:conversionContext:file:line:) + 141 (DatabaseValueConversion.swift:141)
2  Goodnotes                      0x1fb549c partial apply for implicit closure #1 in static DatabaseValueConvertible.decodeIfPresent(from:atUncheckedIndex:) + 4408743068 (<compiler-generated>:4408743068)
3  Goodnotes                      0x1fb4394 fatalConversionError<A>(to:from:in:atColumn:file:line:) + 4408738708 (<compiler-generated>:4408738708)
4  Goodnotes                      0x1be1ae8 specialized FolderToFolderItem.init(row:) + 48 (FolderToFolderItem +GRDB.swift:48)
5  Goodnotes                      0x1bdff4c protocol witness for FetchableRecord.init(row:) in conformance FolderToFolderItem + 4404723532 (<compiler-generated>:4404723532)
6  Goodnotes                      0x207bcb8 RecordCursor.next() + 456 (FetchableRecord.swift:456)
7  Goodnotes                      0x207c9b4 protocol witness for Cursor.next() in conformance RecordCursor<A> + 4409559476 (<compiler-generated>:4409559476)
8  Goodnotes                      0x1f7001c specialized Array.init<A>(_:) + 22 (Cursor.swift:22)
9  Goodnotes                      0x207bb90 static FetchableRecord.fetchAll(_:arguments:adapter:) + 189 (FetchableRecord.swift:189)
10 Goodnotes                      0x207bee8 static FetchableRecord.fetchAll<A>(_:_:) + 334 (FetchableRecord.swift:334)
11 Goodnotes                      0x1d2605c specialized closure #1 in closure #1 in Reactive<A>.fetchAll(in:) + 4406059100 (<compiler-generated>:4406059100)
12 Goodnotes                      0x1d2742c partial apply for specialized closure #1 in closure #1 in Reactive<A>.fetchAll(in:) + 4406064172 (<compiler-generated>:4406064172)
13 Goodnotes                      0x1f9c28c closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408640140
14 Goodnotes                      0x1fa13c8 partial apply for closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660936 (<compiler-generated>:4408660936)
15 Goodnotes                      0x1f8e5f8 Database.inTransaction(_:_:) + 821 (Database.swift:821)
16 Goodnotes                      0x1f9c18c closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408639884
17 Goodnotes                      0x1fa13a8 partial apply for closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660904 (<compiler-generated>:4408660904)
18 Goodnotes                      0x1fe00ac closure #1 in SerializedDatabase.sync<A>(_:) + 4408918188
19 Goodnotes                      0x1fe0e34 partial apply for closure #1 in SerializedDatabase.reentrantSync<A>(_:) + 4408921652
20 Goodnotes                      0x1fe0ee0 partial apply for closure #1 in SerializedDatabase.sync<A>(_:) + 4408921824
21 libswiftDispatch.dylib         0x17f4 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
22 libswiftDispatch.dylib         0x8a90 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
23 libswiftDispatch.dylib         0x297c closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
24 libswiftDispatch.dylib         0x2a90 partial apply for thunk for @callee_guaranteed () -> () + 28
25 libswiftDispatch.dylib         0x25ec thunk for @escaping @callee_guaranteed () -> () + 28
26 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
27 libdispatch.dylib              0x13428 _dispatch_lane_barrier_sync_invoke_and_complete + 56
28 libswiftDispatch.dylib         0x1f38 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 180
29 libswiftDispatch.dylib         0x10dc partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 56
30 libswiftDispatch.dylib         0x1c48 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
31 libswiftDispatch.dylib         0x1188 OS_dispatch_queue.sync<A>(execute:) + 164
32 Goodnotes                      0x1fdf024 SerializedDatabase.sync<A>(_:) + 96 (SerializedDatabase.swift:96)
33 Goodnotes                      0x1fa0a48 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658504
34 Goodnotes                      0x1fa0948 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658248
35 Goodnotes                      0x20a05c0 Pool.get<A>(block:) + 95 (Pool.swift:95)
36 Goodnotes                      0x1f9c7f8 DatabasePool.read<A>(_:) + 4408641528
37 Goodnotes                      0x1f9bcf4 DatabasePool.read<A>(_:) + 4408638708
38 Goodnotes                      0x1f9d6a0 protocol witness for DatabaseReader.read<A>(_:) in conformance DatabasePool + 4408645280 (<compiler-generated>:4408645280)
39 Goodnotes                      0x1c97a68 LoggingDatabaseWriter.read<A>(_:) + 127 (LoggingDatabaseWriter.swift:127)
40 Goodnotes                      0x1c982f4 protocol witness for DatabaseReader.read<A>(_:) in conformance LoggingDatabaseWriter + 4405478132 (<compiler-generated>:4405478132)
41 Goodnotes                      0x1d21c90 specialized closure #1 in Reactive<A>.fetchAll(in:) + 638 (LibraryManager.swift:638)
42 Goodnotes                      0x19da13c closure #1 in static PrimitiveSequenceType<>.create(subscribe:) + 4402602300
43 Goodnotes                      0x1991de0 AnonymousObservableSink.run(_:) + 60 (Create.swift:60)
44 Goodnotes                      0x1991f84 AnonymousObservable.run<A>(_:cancel:) + 76 (Create.swift:76)
45 Goodnotes                      0x19c6688 closure #1 in Producer.subscribe<A>(_:) + 4402521736
46 Goodnotes                      0x19c5d9c specialized CurrentThreadScheduler.schedule<A>(_:action:) + 18 (CurrentThreadScheduler.swift:18)
47 Goodnotes                      0x19c5c9c Producer.subscribe<A>(_:) + 24 (Producer.swift:24)
48 Goodnotes                      0x19ba7dc protocol witness for ObservableType.subscribe<A>(_:) in conformance Observable<A> + 4402472924 (<compiler-generated>:4402472924)
49 Goodnotes                      0x19dfe74 closure #1 in SubscribeOnSink.run() + 4402626164
50 Goodnotes                      0x199b93c closure #1 in DispatchQueueConfiguration.schedule<A>(_:action:) + 4402346300
51 Goodnotes                      0x19918c0 thunk for @escaping @callee_guaranteed () -> () + 4402305216 (<compiler-generated>:4402305216)
52 libdispatch.dylib              0x2320 _dispatch_call_block_and_release + 32
53 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
54 libdispatch.dylib              0x7330 _dispatch_continuation_pop + 504
55 libdispatch.dylib              0x6998 _dispatch_async_redirect_invoke + 584
56 libdispatch.dylib              0x15944 _dispatch_root_queue_drain + 396
57 libdispatch.dylib              0x16158 _dispatch_worker_thread2 + 164
58 libsystem_pthread.dylib        0xda0 _pthread_wqthread + 228
59 libsystem_pthread.dylib        0xb7c start_wqthread + 8

GRDB.DatabasePool.reader.1
0  libsystem_kernel.dylib         0x1dbc pread + 8
1  libsqlite3.dylib               0x8e228 sqlite3_sourceid + 7616
2  libsqlite3.dylib               0x7078 sqlite3_snprintf + 7904
3  libsqlite3.dylib               0x19e60 sqlite3_log + 3448
4  libsqlite3.dylib               0x97ca8 sqlite3_sourceid + 47168
5  libsqlite3.dylib               0x5b014 sqlite3_close + 1032
6  libsqlite3.dylib               0x40634 sqlite3_get_table + 11376
7  libsqlite3.dylib               0x35e24 sqlite3_step + 44104
8  libsqlite3.dylib               0x2b5b4 sqlite3_step + 984
9  Goodnotes                      0x207bc60 RecordCursor.next() + 451 (FetchableRecord.swift:451)
10 Goodnotes                      0x207c9b4 protocol witness for Cursor.next() in conformance RecordCursor<A> + 4409559476 (<compiler-generated>:4409559476)
11 Goodnotes                      0x1f7009c specialized Array.init<A>(_:) + 22 (Cursor.swift:22)
12 Goodnotes                      0x207bb90 static FetchableRecord.fetchAll(_:arguments:adapter:) + 189 (FetchableRecord.swift:189)
13 Goodnotes                      0x207bee8 static FetchableRecord.fetchAll<A>(_:_:) + 334 (FetchableRecord.swift:334)
14 Goodnotes                      0x207b980 FetchRequest<>.fetchAll(_:) + 397 (FetchableRecord.swift:397)
15 Goodnotes                      0x1e8f348 closure #1 in GRDBSyncStore.listAssetItems(stream:needsSyncOnly:) + 4407538504
16 Goodnotes                      0x1e9437c partial apply for closure #1 in GRDBSyncStore.listAssetItems(stream:needsSyncOnly:) + 4407559036 (<compiler-generated>:4407559036)
17 Goodnotes                      0x1f9c28c closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408640140
18 Goodnotes                      0x1fa13c8 partial apply for closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660936 (<compiler-generated>:4408660936)
19 Goodnotes                      0x1f8e5f8 Database.inTransaction(_:_:) + 821 (Database.swift:821)
20 Goodnotes                      0x1f9c18c closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408639884
21 Goodnotes                      0x1fa13a8 partial apply for closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660904 (<compiler-generated>:4408660904)
22 Goodnotes                      0x1fe00ac closure #1 in SerializedDatabase.sync<A>(_:) + 4408918188
23 Goodnotes                      0x1fe0e34 partial apply for closure #1 in SerializedDatabase.reentrantSync<A>(_:) + 4408921652
24 Goodnotes                      0x1fe0ee0 partial apply for closure #1 in SerializedDatabase.sync<A>(_:) + 4408921824
25 libswiftDispatch.dylib         0x17f4 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
26 libswiftDispatch.dylib         0x8a90 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
27 libswiftDispatch.dylib         0x297c closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
28 libswiftDispatch.dylib         0x2a90 partial apply for thunk for @callee_guaranteed () -> () + 28
29 libswiftDispatch.dylib         0x25ec thunk for @escaping @callee_guaranteed () -> () + 28
30 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
31 libdispatch.dylib              0x13428 _dispatch_lane_barrier_sync_invoke_and_complete + 56
32 libswiftDispatch.dylib         0x1f38 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 180
33 libswiftDispatch.dylib         0x10dc partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 56
34 libswiftDispatch.dylib         0x1c48 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
35 libswiftDispatch.dylib         0x1188 OS_dispatch_queue.sync<A>(execute:) + 164
36 Goodnotes                      0x1fdf024 SerializedDatabase.sync<A>(_:) + 96 (SerializedDatabase.swift:96)
37 Goodnotes                      0x1fa0a48 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658504
38 Goodnotes                      0x1fa0948 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658248
39 Goodnotes                      0x20a05c0 Pool.get<A>(block:) + 95 (Pool.swift:95)
40 Goodnotes                      0x1f9c7f8 DatabasePool.read<A>(_:) + 4408641528
41 Goodnotes                      0x1f9bcf4 DatabasePool.read<A>(_:) + 4408638708
42 Goodnotes                      0x1f9d6a0 protocol witness for DatabaseReader.read<A>(_:) in conformance DatabasePool + 4408645280 (<compiler-generated>:4408645280)
43 Goodnotes                      0x1e8f034 GRDBSyncStore.listAssetItems(stream:needsSyncOnly:) + 72 (GRDBSyncStore.swift:72)
44 Goodnotes                      0x1e914c4 protocol witness for SyncStore.listAssetItems(stream:needsSyncOnly:) in conformance GRDBSyncStore<A, B> + 4407547076 (<compiler-generated>:4407547076)
45 Goodnotes                      0x1b3f30c partial apply for specialized implicit closure #10 in implicit closure #9 in AnySyncStore.init<A>(base:) + 4404065036
46 Goodnotes                      0x1b3fe84 partial apply for specialized implicit closure #10 in implicit closure #9 in AnySyncStore.init<A>(base:) + 4404067972
47 Goodnotes                      0x1ea3340 AnySyncStore.listAssetItems(stream:needsSyncOnly:) + 156 (SyncStore.swift:156)
48 Goodnotes                      0x1d8267c closure #1 in SyncManagerV2.listItemsToSync() + 4406437500
49 Goodnotes                      0x1d7ad9c SyncManagerV2.listItemsToSync() + 313 (SyncManagerV2.swift:313)
50 Goodnotes                      0x1d87058 partial apply for closure #1 in SyncManagerV2.start() + 88 (SyncManagerV2.swift:88)
51 Goodnotes                      0x1d8709c partial apply for thunk for @escaping @callee_guaranteed () -> (@owned [EventStreamIndex : (stream: SyncStreamItem<A, B>, assets: [SyncAssetItem<A, B>])], @error @owned Error) + 4406456476 (<compiler-generated>:4406456476)
52 Goodnotes                      0x21cc444 closure #1 in closure #1 in Thenable.map<A>(on:flags:_:) + 4410934340
53 Goodnotes                      0x21c3528 thunk for @escaping @callee_guaranteed () -> () + 4410897704 (<compiler-generated>:4410897704)
54 libdispatch.dylib              0x2320 _dispatch_call_block_and_release + 32
55 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
56 libdispatch.dylib              0xb534 _dispatch_lane_serial_drain + 668
57 libdispatch.dylib              0xc0a4 _dispatch_lane_invoke + 384
58 libdispatch.dylib              0x16cdc _dispatch_workloop_worker_thread + 648
59 libsystem_pthread.dylib        0xddc _pthread_wqthread + 288
60 libsystem_pthread.dylib        0xb7c start_wqthread + 8


GRDB.DatabasePool.reader.1
0   libsystem_kernel.dylib         0x1dbc pread + 8
1   libsqlite3.dylib               0x8e228 sqlite3_sourceid + 7616
2   libsqlite3.dylib               0x7078 sqlite3_snprintf + 7904
3   libsqlite3.dylib               0x19e60 sqlite3_log + 3448
4   libsqlite3.dylib               0x97ca8 sqlite3_sourceid + 47168
5   libsqlite3.dylib               0xa3388 sqlite3_sourceid + 93984
6   libsqlite3.dylib               0x37528 sqlite3_step + 49996
7   libsqlite3.dylib               0xa02a4 sqlite3_sourceid + 81468
8   libsqlite3.dylib               0x9f3ec sqlite3_sourceid + 77700
9   libsqlite3.dylib               0x35974 sqlite3_step + 42904
10  libsqlite3.dylib               0x2b5b4 sqlite3_step + 984
11  Goodnotes                      0x1fcf358 specialized static Row.fetchAll(_:arguments:adapter:) + 799 (Row.swift:799)
12  Goodnotes                      0x1fc3fd4 static Row.fetchAll<A>(_:_:) + 1041 (Row.swift:1041)
13  Goodnotes                      0x1d30930 specialized ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(_:documentIds:dueTimestamp:excludeUnsubscribed:) + 179 (ListSrsReviewStates.swift:179)
14  Goodnotes                      0x1d2fbbc closure #1 in closure #1 in ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(folderID:dueTimestamp:excludeUnsubscribed:onChangeHandler:onErrorHandler:) + 4406098876
15  Goodnotes                      0x1d310f0 partial apply for closure #1 in closure #1 in ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(folderID:dueTimestamp:excludeUnsubscribed:onChangeHandler:onErrorHandler:) + 4406104304 (<compiler-generated>:4406104304)
16  Goodnotes                      0x1f9c28c closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408640140
17  Goodnotes                      0x1fa13c8 partial apply for closure #1 in closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660936 (<compiler-generated>:4408660936)
18  Goodnotes                      0x1f8e5f8 Database.inTransaction(_:_:) + 821 (Database.swift:821)
19  Goodnotes                      0x1f9c18c closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408639884
20  Goodnotes                      0x1fa13a8 partial apply for closure #1 in closure #1 in DatabasePool.read<A>(_:) + 4408660904 (<compiler-generated>:4408660904)
21  Goodnotes                      0x1fe00ac closure #1 in SerializedDatabase.sync<A>(_:) + 4408918188
22  Goodnotes                      0x1fe0e34 partial apply for closure #1 in SerializedDatabase.reentrantSync<A>(_:) + 4408921652
23  Goodnotes                      0x1fe0ee0 partial apply for closure #1 in SerializedDatabase.sync<A>(_:) + 4408921824
24  libswiftDispatch.dylib         0x17f4 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
25  libswiftDispatch.dylib         0x8a90 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
26  libswiftDispatch.dylib         0x297c closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
27  libswiftDispatch.dylib         0x2a90 partial apply for thunk for @callee_guaranteed () -> () + 28
28  libswiftDispatch.dylib         0x25ec thunk for @escaping @callee_guaranteed () -> () + 28
29  libdispatch.dylib              0x3eac _dispatch_client_callout + 20
30  libdispatch.dylib              0x13428 _dispatch_lane_barrier_sync_invoke_and_complete + 56
31  libswiftDispatch.dylib         0x1f38 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 180
32  libswiftDispatch.dylib         0x10dc partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 56
33  libswiftDispatch.dylib         0x1c48 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
34  libswiftDispatch.dylib         0x1188 OS_dispatch_queue.sync<A>(execute:) + 164
35  Goodnotes                      0x1fdf024 SerializedDatabase.sync<A>(_:) + 96 (SerializedDatabase.swift:96)
36  Goodnotes                      0x1fa0a48 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658504
37  Goodnotes                      0x1fa0948 partial apply for closure #1 in DatabasePool.read<A>(_:) + 4408658248
38  Goodnotes                      0x20a05c0 Pool.get<A>(block:) + 95 (Pool.swift:95)
39  Goodnotes                      0x1f9c7f8 DatabasePool.read<A>(_:) + 4408641528
40  Goodnotes                      0x1f9bcf4 DatabasePool.read<A>(_:) + 4408638708
41  Goodnotes                      0x1f9d6a0 protocol witness for DatabaseReader.read<A>(_:) in conformance DatabasePool + 4408645280 (<compiler-generated>:4408645280)
42  Goodnotes                      0x1c97a68 LoggingDatabaseWriter.read<A>(_:) + 127 (LoggingDatabaseWriter.swift:127)
43  Goodnotes                      0x1c982f4 protocol witness for DatabaseReader.read<A>(_:) in conformance LoggingDatabaseWriter + 4405478132 (<compiler-generated>:4405478132)
44  Goodnotes                      0x1d2fb0c closure #1 in ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(folderID:dueTimestamp:excludeUnsubscribed:onChangeHandler:onErrorHandler:) + 4406098700
45  Goodnotes                      0x1d3103c partial apply for closure #1 in ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(folderID:dueTimestamp:excludeUnsubscribed:onChangeHandler:onErrorHandler:) + 4406104124 (<compiler-generated>:4406104124)
46  Goodnotes                      0x19af614 MapSink.on(_:) + 42 (Map.swift:42)
47  Goodnotes                      0x19af80c protocol witness for ObserverType.on(_:) in conformance MapSink<A, B> + 4402427916 (<compiler-generated>:4402427916)
48  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
49  Goodnotes                      0x196fa40 AsSingleSink.on(_:) + 28 (AsSingle.swift:28)
50  Goodnotes                      0x196fc0c protocol witness for ObserverType.on(_:) in conformance AsSingleSink<A> + 4402166796 (<compiler-generated>:4402166796)
51  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
52  Goodnotes                      0x19e4450 TakeCountSink.on(_:) + 87 (Take.swift:87)
53  Goodnotes                      0x19e4550 protocol witness for ObserverType.on(_:) in conformance TakeCountSink<A> + 4402644304 (<compiler-generated>:4402644304)
54  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
55  Goodnotes                      0x19af6ec MapSink.on(_:) + 43 (Map.swift:43)
56  Goodnotes                      0x19af80c protocol witness for ObserverType.on(_:) in conformance MapSink<A, B> + 4402427916 (<compiler-generated>:4402427916)
57  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
58  Goodnotes                      0x199d2b4 DistinctUntilChangedSink.on(_:) + 105 (DistinctUntilChanged.swift:105)
59  Goodnotes                      0x199d504 protocol witness for ObserverType.on(_:) in conformance DistinctUntilChangedSink<A, B> + 4402353412 (<compiler-generated>:4402353412)
60  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
61  Goodnotes                      0x19af6ec MapSink.on(_:) + 43 (Map.swift:43)
62  Goodnotes                      0x19af80c protocol witness for ObserverType.on(_:) in conformance MapSink<A, B> + 4402427916 (<compiler-generated>:4402427916)
63  Goodnotes                      0x19d820c closure #1 in ShareReplay1WhileConnectedConnection.synchronized_subscribe<A>(_:) + 4402594316
64  Goodnotes                      0x19d86dc ShareReplay1WhileConnected.subscribe<A>(_:) + 20 (Lock.swift:20)
65  Goodnotes                      0x19af92c Map.run<A>(_:cancel:) + 74 (Map.swift:74)
66  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
67  Goodnotes                      0x199d618 DistinctUntilChanged.run<A>(_:cancel:) + 135 (DistinctUntilChanged.swift:135)
68  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
69  Goodnotes                      0x19af92c Map.run<A>(_:cancel:) + 74 (Map.swift:74)
70  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
71  Goodnotes                      0x19e46c8 TakeCount.run<A>(_:cancel:) + 117 (Take.swift:117)
72  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
73  Goodnotes                      0x196fd40 AsSingle.run<A>(_:cancel:) + 49 (AsSingle.swift:49)
74  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
75  Goodnotes                      0x19af92c Map.run<A>(_:cancel:) + 74 (Map.swift:74)
76  Goodnotes                      0x19c5bc8 Producer.subscribe<A>(_:) + 18 (Producer.swift:18)
77  Goodnotes                      0x19bad68 ObservableType.subscribe(_:) + 24 (ObservableType+Extensions.swift:24)
78  Goodnotes                      0x19da4a4 PrimitiveSequenceType<>.subscribe(_:) + 53 (Single.swift:53)
79  Goodnotes                      0x19d9e84 PrimitiveSequenceType<>.subscribe(onSuccess:onFailure:onDisposed:) + 161 (Single.swift:161)
80  Goodnotes                      0x1d2f0b0 ListSrsReviewStatesSubscriptionHandler.listSrsReviewStates(folderID:dueTimestamp:excludeUnsubscribed:onChangeHandler:onErrorHandler:) + 95 (ListSrsReviewStates.swift:95)
81  Goodnotes                      0x1d2ebe8 implicit closure #2 in implicit closure #1 in ListSrsReviewStatesSubscriptionHandler.register(to:) + 45 (ListSrsReviewStates.swift:45)
82  Goodnotes                      0x1d31108 partial apply for implicit closure #2 in implicit closure #1 in ListSrsReviewStatesSubscriptionHandler.register(to:) + 4406104328 (<compiler-generated>:4406104328)
83  Goodnotes                      0x1d2edcc thunk for @escaping @callee_guaranteed (@guaranteed ListSrsReviewStates.Request, @guaranteed @escaping @callee_guaranteed (@guaranteed ListSrsReviewStates.Response) -> (), @guaranteed @escaping @callee_guaranteed (@guaranteed Error) -> ()) -> (@out Unsubscribable, @error @owned Error) + 4406095308 (<compiler-generated>:4406095308)
84  Goodnotes                      0x1d31120 partial apply for thunk for @escaping @callee_guaranteed (@guaranteed ListSrsReviewStates.Request, @guaranteed @escaping @callee_guaranteed (@guaranteed ListSrsReviewStates.Response) -> (), @guaranteed @escaping @callee_guaranteed (@guaranteed Error) -> ()) -> (@out Unsubscribable, @error @owned Error) + 4406104352 (<compiler-generated>:4406104352)
85  Goodnotes                      0x1a049fc thunk for @escaping @callee_guaranteed (@in_guaranteed A.Subscription.Request, @guaranteed @escaping @callee_guaranteed @substituted <A> (@in_guaranteed A) -> () for <A.Subscription.Response>, @guaranteed @escaping @callee_guaranteed (@guaranteed Error) -> ()) -> (@out Unsubscribable, @error @owned Error) + 4402776572 (<compiler-generated>:4402776572)
86  Goodnotes                      0x1a0686c partial apply for thunk for @escaping @callee_guaranteed (@in_guaranteed A.Subscription.Request, @guaranteed @escaping @callee_guaranteed @substituted <A> (@in_guaranteed A) -> () for <A.Subscription.Response>, @guaranteed @escaping @callee_guaranteed (@guaranteed Error) -> ()) -> (@out Unsubscribable, @error @owned Error) + 4402784364 (<compiler-generated>:4402784364)
87  Goodnotes                      0x1a05e34 Bus.subscribe<A>(_:request:onChangeHandler:onErrorHandler:) + 165 (Bus.swift:165)
88  Goodnotes                      0x1b16fd4 closure #1 in closure #1 in GoodNotesService.subscribe<A>(_:request:) + 4403900372
89  Goodnotes                      0x1991de0 AnonymousObservableSink.run(_:) + 60 (Create.swift:60)
90  Goodnotes                      0x1991f84 AnonymousObservable.run<A>(_:cancel:) + 76 (Create.swift:76)
91  Goodnotes                      0x19c6688 closure #1 in Producer.subscribe<A>(_:) + 4402521736
92  Goodnotes                      0x19c5d9c specialized CurrentThreadScheduler.schedule<A>(_:action:) + 18 (CurrentThreadScheduler.swift:18)
93  Goodnotes                      0x19c5c9c Producer.subscribe<A>(_:) + 24 (Producer.swift:24)
94  Goodnotes                      0x19b4370 MergeSink.subscribeInner(_:) + 480 (Merge.swift:480)
95  Goodnotes                      0x19b418c MergeSink.on(_:) + 459 (Merge.swift:459)
96  Goodnotes                      0x19b5104 protocol witness for ObserverType.on(_:) in conformance MergeSink<A, B, C> + 4402450692 (<compiler-generated>:4402450692)
97  Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
98  Goodnotes                      0x196fa40 AsSingleSink.on(_:) + 28 (AsSingle.swift:28)
99  Goodnotes                      0x196fc0c protocol witness for ObserverType.on(_:) in conformance AsSingleSink<A> + 4402166796 (<compiler-generated>:4402166796)
100 Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
101 Goodnotes                      0x19e4450 TakeCountSink.on(_:) + 87 (Take.swift:87)
102 Goodnotes                      0x19e4550 protocol witness for ObserverType.on(_:) in conformance TakeCountSink<A> + 4402644304 (<compiler-generated>:4402644304)
103 Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
104 Goodnotes                      0x198aca4 CompactMapSink.on(_:) + 42 (CompactMap.swift:42)
105 Goodnotes                      0x198ada4 protocol witness for ObserverType.on(_:) in conformance CompactMapSink<A, B> + 4402277796 (<compiler-generated>:4402277796)
106 Goodnotes                      0x19dcf84 Sink.forwardOn(_:) + 35 (Sink.swift:35)
107 Goodnotes                      0x19dfba4 SubscribeOnSink.on(_:) + 66 (SubscribeOn.swift:66)
108 Goodnotes                      0x19e001c protocol witness for ObserverType.on(_:) in conformance SubscribeOnSink<A, B> + 4402626588 (<compiler-generated>:4402626588)
109 Goodnotes                      0x19762f8 BehaviorSubject.synchronized_subscribe<A>(_:) + 121 (BehaviorSubject.swift:121)
110 Goodnotes                      0x1976014 BehaviorSubject.subscribe<A>(_:) + 20 (Lock.swift:20)
111 Goodnotes                      0x19ba7dc protocol witness for ObservableType.subscribe<A>(_:) in conformance Observable<A> + 4402472924 (<compiler-generated>:4402472924)
112 Goodnotes                      0x19dfe74 closure #1 in SubscribeOnSink.run() + 4402626164
113 Goodnotes                      0x199b93c closure #1 in DispatchQueueConfiguration.schedule<A>(_:action:) + 4402346300
114 Goodnotes                      0x19918c0 thunk for @escaping @callee_guaranteed () -> () + 4402305216 (<compiler-generated>:4402305216)
115 libdispatch.dylib              0x2320 _dispatch_call_block_and_release + 32
116 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
117 libdispatch.dylib              0xb534 _dispatch_lane_serial_drain + 668
118 libdispatch.dylib              0xc0a4 _dispatch_lane_invoke + 384
119 libdispatch.dylib              0x16cdc _dispatch_workloop_worker_thread + 648
120 libsystem_pthread.dylib        0xddc _pthread_wqthread + 288
121 libsystem_pthread.dylib        0xb7c start_wqthread + 8
```

**Comment by groue** [comment]
Thanks for the heads up, @0xpablo 

> We only use a single `DatabaseQueue`.

I see `DatabasePool` in the crash logs.

> ```swift
> // Workaround GRDB bug fixed in 5.8.0
> // TODO: remove once GRDB has been upgraded.
> try dbQueue.inDatabase { $0.clearSchemaCache() }
> ```

This won't help with a DatabasePool, because it maintains several SQLite connection. The above sample code only clears the cache for the writer connection, and leaves reader connections intact. We see in the crash logs that they all happen from readers (`GRDB.DatabasePool.reader.1`, etc).

> The associations part is interesting, I took a look at our observation and while we don't use the including API because we build the query ourselves, we have quite a bunch of LEFT JOIN statements in this case. I suppose the including API uses JOIN behind the scenes so this might be related.

If this is accurate (your app does not use `including`), then the GRDB code that relies on an accurate schema cache is not used.

Maybe we're on a wrong track focusing the schema cache.

**Comment by 0xpablo** [comment]
Oh that's very interesting, yes, we are indeed using a `DatabasePool`, I had to change the code to this:
```swift
try container.databasePool.write { db in
    db.clearSchemaCache()
}
```
But I understand now this will not do anything on the readers, so the workaround won't do much as you mention.

I checked again and I don't think we are using the `including` APIs, but I added breakpoints to the Schema Cache after the initialization, to see if it was used in the queries that we do and it looks like is used when we use fetchAll.

<img width="478" alt="image" src="https://github.com/groue/GRDB.swift/assets/889020/9a7c6558-7ce9-4ad2-8598-2de1b6eb240f">
<img width="513" alt="image" src="https://github.com/groue/GRDB.swift/assets/889020/58abcc0b-b322-4f8f-af48-c3c30797da96">

To apply the workaround to readers as well there is a convenient `forEachConnection`, but it's `fileprivate` so we cannot use it. Will `invalidateReadOnlyConnections` achieve the same effect? (after clearing the cache for the writer connection).

Thank you!

**Comment by groue** [comment]
> To apply the workaround to readers as well there is a convenient `forEachConnection`, but it's `fileprivate` so we cannot use it. Will `invalidateReadOnlyConnections` achieve the same effect? I think so, as I assume that will cause the pool to call the `makeElement` closure, which returns a brand new cache.

Yes, `invalidateReadOnlyConnections` can help.

`invalidateReadOnlyConnections` does not close *currently used* connections. For example, calling this method does not make eventual concurrent reads fail. It just makes sure that currently used connections won't be used again, and force all subsequent reads to open new connection(s).

I don't know what is the level of concurrency in your app, and whether or not it is possible that some read accesses are performed on the database *before* the migrations have completed. I'm not saying your app is doing so, but I've seen enough apps in my life to know that abstraction layers can pile up and make it pretty much impossible to answer such a simple question with 100% accuracy 😅

**Comment by 0xpablo** [comment]
I can confirm that using `invalidateReadOnlyConnections` as the last step of our initialization process (after performing the migrations) fixed our issue! We saw no more crashes in a version we pushed with the fix. 

Thanks again for your support!

**Comment by groue** [comment]
You're welcome! Happy GRDB!
